sympy/matrices/common.py

```diff
@@ -732,32 +756,226 @@ def size(m):
-        rows = kwargs.get('rows', diag_rows)
-        cols = kwargs.get('cols', diag_cols)
+        def minsize(d):
+            size = d.pop('size', None)
+            r = max(0, -min(d)) + 1
+            c = -min(0, -max(d)) + 1
+            for k in d:
+                if not type(d[k]) is list:
+                    continue
+                L = len(d[k])
+                r = max(r, L if k > 0 else abs(k) + L)
+                c = max(c, L if k < 0 else abs(k) + L)
+            if size:
+                R, C = size
+                if r > R or c > C:
+                    raise ValueError(filldedent('''
+                        the dict size is too small
+                        for the given elements'''))
+                r, c = R, C
+                d['size'] = size
+            return r, c
+
+        def arg_size(args):
+            prev = None
+            max_row, max_col, o_r, o_c, g_r, g_c = [0]*6
+            for a_ in args:
+                a_size = size(a_)
+                if type(a_) is dict and ('goto' in a_ or 'move' in a_):
+                    prev_size = size(a_)
+                    if 'goto' in a_:
+                        prev = 'goto'
+                        g_r, g_c = a_size
+                    else:
+                        prev = 'move'
+                else:
+                    o_r += a_size[0]
+                    o_c += a_size[1]
+                    if prev is None:
+                        max_row, max_col = o_r, o_c
+                    elif 'goto' in prev:
+                        g_r += a_size[0]
+                        g_c += a_size[1]
+                        max_row = max(max_row, g_r)
+                        max_col = max(max_col, g_c)
+                    elif 'move' in prev:
+                        max_row = max(max_row, o_r - prev_size[0])
+                        max_col = max(max_col, o_c - prev_size[1])
+            return max_row, max_col
+
+        def standardize_values(d):
+            for k, value in d.items():
+                if k == 'size':
+                    continue
+                as_int(k)  # check
+                if isinstance(value, list):
+                    continue
+                elif isfunction(value):
+                    num_args = _getnargs(value)
+                    if num_args not in (1, 2):
+                        raise ValueError(filldedent('''
+                            The functions in dict-described
+                            diagonals must have 1 or 2 args.'''))
+                else:
+                    value = sympify(value)
+                    if isinstance(value, Expr
+                        ) and not hasattr(value, 'rows'):
+                        d[k] = value
+                    else:
+                        raise TypeError(filldedent(
+                            '''expecting list, function
+                            or Expr in dict'''))
+
+        def kern(d):
+            if 'move' in d or 'goto' in d:
```

Comment:
--> ```if type(d) is dict and ('move' in d or 'goto' in d):```

Reply:
Not required.. I guess cause we use `kern(d)` only when d is a dict.